### ISSUE TITLE
salfter's switchers custom fan management conflicting with tempcontrol fan settings

### ISSUE THREAD

**Issue by papampi** [papampi/nvOC_by_fullzero#58]
Getting this error with mph_switch, any idea why?

```
m1@m1-desktop-100:~$ python2.7 mph_switch mph_conf.json 
Traceback (most recent call last):
  File "mph_switch", line 14, in <module>
    cfg=json.loads(open(sys.argv[1]).read())
  File "/usr/lib/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded

```

mph_conf.json:
```
{
  "output_dir": "/home/m1/",
  "user_name": "papampi",
  "miner_name": "nv100",
  "card_type": "nvidia",
  "currency": "USD",
  "pwrcost": "0.01",
  "min_profit": "1",
  "miners":
  {
    "Blake-Vanilla":
    {
      "bin": "/home/m1/miners/SPccminer/ccminer -a vanilla -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 12.3, "power": 0.389
    },
    "Cryptonight":
    {
      "bin": "/home/m1/miners/KTccminer-cryptonight/ccminer -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.000002, "power": 0.325
    },
    "Equihash":
    {
      "bin": "/home/m1/miners/ewbf/v3_4/miner --eexit 3 --fee 0 --pec --server {HOST} --user {NAME}.{MINER} --pass z --port {PORT}",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.000001140, "power": 0.397
    },
    "Ethash":
    {
      "bin": "/home/m1/miners/ethminer/latest/ethminer -S {HOST}:{PORT} -O {NAME}.{MINER}:x -U",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.087, "power": 0.397
    },
    "Groestl":
    {
      "bin": "/home/m1/miners/SPccminer/ccminer -a groestl -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.0853, "power": 0.392
    },
    "Keccak":
    {
      "bin": "/home/m1/miners/SPccminer/ccminer -a keccak -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 1.45, "power": 0.391
    },
    "Lyra2RE2":
    {
      "bin": "/home/m1/miners/SPccminer/ccminer -a lyra2v2 -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.081, "power": 0.386
    },
    "Myriad-Groestl":
    {
      "bin": "/home/m1/miners/SPccminer/ccminer -a myr-gr -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": , "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.158, "power": 0.391
    },
    "NeoScrypt":
    {
      "bin": "/home/m1/miners/KTccminer/ccminer -a neoscrypt -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.002160, "power": 0.397
    },
    "Qubit":
    {
      "bin": "/home/m1/miners/SPccminer/ccminer -a qubit -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.0509, "power": 0.382
    },
    "Scrypt":
    {
      "bin": "/home/m1/SPccminer/ccminer -a scrypt -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.00224, "power": 0.403
    },
    "Sia":
    {
      "bin": "/home/m1/miners/SPccminer/ccminer -a sia -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 3.700, "power": 0.403
    },
    "Skein":
    {
      "bin": "/home/m1/miners/ASccminer/ccminer -a skein -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.835, "power": 0.383
    },
    "X11":
    {
      "bin": "/home/m1/miners/TPccminer/ccminer -a x11 -o stratum+tcp://{HOST}:{PORT} -u {NAME}.{MINER} -p x",
      "power_limit": [100,78,78], "gpu_oc": 100, "mem_oc": 100, "fan": 0,
      "speed": 0.0327, "power": 0.403
    }
  }
}

```

**Comment by papampi** [comment]
Removed "Blake-Vanilla" and changed "Cryptonight" to "Cryptonight-Monero", still same error.

**Comment by LuKePicci** [comment]
Power_limit is missing for myriad-groestl

________________________________
Da: Payam Nab <notifications@github.com>
Inviato: Saturday, May 12, 2018 9:19:21 AM
A: papampi/nvOC_by_fullzero_Community_Release
Cc: Subscribed
Oggetto: Re: [papampi/nvOC_by_fullzero_Community_Release] mph_switch error ... (#58)


Removed "Blake-Vanilla" and changed "Monero" to "Cryptonight-Monero", still same error.

**Comment by papampi** [comment]
Your eagle eyes ... 
Fixed it in 3main, also needed to change the output_dir to :
https://github.com/papampi/nvOC_by_fullzero_Community_Release/blob/cb6c236dc5291feb235b4aa88567abae1b025062/3main#L676


But temp control giving error, no problem on manual coin

```
GPU 0, Sat May 12 13:37:10 +0430 2018 - Adjusting fan from: 30 to: 50 Temp: 58


ERROR: Error assigning value 50 to attribute 'GPUTargetFanSpeed'
       (m1-desktop-100:0[fan:0]) as specified in assignment
       '[fan:0]/GPUTargetFanSpeed=50' (Unknown Error).

GPU 1, Target temp: 65, Current: 48, Diff: 17, Fan: 53, Power: 78.10

GPU 1, Sat May 12 13:37:11 +0430 2018 - Adjusting fan from: 53 to: 50 Temp: 48


ERROR: Error assigning value 50 to attribute 'GPUTargetFanSpeed'
       (m1-desktop-100:0[fan:1]) as specified in assignment
       '[fan:1]/GPUTargetFanSpeed=50' (Unknown Error).

GPU 2, Target temp: 65, Current: 51, Diff: 14, Fan: 64, Power: 77.79

GPU 2, Sat May 12 13:37:11 +0430 2018 - Adjusting fan from: 64 to: 59 Temp: 51


ERROR: Error assigning value 59 to attribute 'GPUTargetFanSpeed'
       (m1-desktop-100:0[fan:2]) as specified in assignment
       '[fan:2]/GPUTargetFanSpeed=59' (Unknown Error).


```

**Comment by LuKePicci** [comment]
Add that missing slashes directly into every python script instead of 3main, such that it's more robust since the standard way of specifying paths is without the ending slash, eventually, a double slash would occur it will also work fine.

**Comment by LuKePicci** [comment]
When you say the tempcontrol is fine in manual coin, does it also mean that is fine with WTM switcher?

**Comment by papampi** [comment]
Like this?
log=open(outputDir + "/" + "current-profit", "w")
or this:
log=open(outputDir/ + "current-profit", "w")

And yes, tested WTM and its working fine with it too.

**Comment by LuKePicci** [comment]
log=open(outputDir + "/current-profit", "w")

**Comment by LuKePicci** [comment]
Ok, I'm trying to figure out why mph and other salfter's switchers are doing things in a certain manner. Will do some changes later today. meanwhile go ahead with #57

**Comment by papampi** [comment]
Looks like there is a fan setting at the end of mph_switch and it conflict with tempcontrol

**Comment by LuKePicci** [comment]
Yeah, but the point here is to understand why the switcher was taking care of managing OC and fans by itself given it's not implemented as a standalone parallel job. All the things I see should make sense only if its developer originally intended to switch the mining algo in a transparent way without the need to loop directly into 3main. Moreover, why the json is created only if auto_start_miner = YES an the guake tab is created twice in LOCAL mode?

**Comment by papampi** [comment]
I never used safter switchers so never checked/edited them, was just checking it for 2.1 and got the temp error.
We can move if auto_start_miner == yes to after creating json, but no idea why it creates 2 guake tabs.
In 3main I see 2 guake commands, one has "-n 1" arg to create new tab, other one dont.
And I was testing it from remote ssh, but my test rig setting is LOCAL.
Will test more when I get home beside my test rig.

**Comment by papampi** [comment]
"Manual_fan == yes && tempcontrol == no ", does not set fan speed too when using salfter switcher

**Comment by papampi** [comment]
Removed the last fan part and all is good for now

**Comment by LuKePicci** [comment]
That one in the python switcher code?

**Comment by LuKePicci** [comment]
Ok, that kind of incompatibilities were expected to emerge since the old 3main control flow was a complete mess. Now, the correct (but annoying) way of fixing this incompatibility should be to manage the manual fan control as like as OC & PL, but I think removing fan management from those switchers could be ok since you're not interested into ALGO_SPECIFIC manual fan policies

**Comment by papampi** [comment]
Yes, now I see that programmatic has those disabled too

**Comment by LuKePicci** [comment]
Fixed into #59
